File: sdk/typescript/src/api/payments.ts
import type { HttpClient, RequestOptions } from "../http.js";

export type CoinbasePlan = "BASIC" | "PRO";
export type CoinbaseInterval = "annual";
export type CreditGateway = "stripe" | "coinbase";
export type StripePlan = "BASIC_MONTHLY" | "BASIC_YEARLY" | "PRO_MONTHLY" | "PRO_YEARLY";

export interface CreateCoinbaseChargeBody {
  /** Subscription plan (BASIC: $10/month, PRO: $25/month). */
  plan: CoinbasePlan;
  /** Billing interval (crypto payments are annual-only). */
  interval?: CoinbaseInterval;
  metadata?: Record<string, unknown>;
}

export interface GetCoinbaseChargeQuery {
  /** Sync the charge status with Coinbase Commerce before returning. */
  sync?: boolean;
}

export interface UpdateAutoRechargeBody {
  enabled?: boolean;
  thresholdUsd?: number;
  rechargeAmountUsd?: number;
}

export interface UpdateAutoRechargeCardBody {
  /** Mark this card as the default payment method. */
  isDefault?: boolean;
  billingDetails?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface CreateCreditTopUpBody {
  /** Amount in USD to add as credits (capped by MAX_TOPUP_USD). */
  amountUsd: number;
  /** Payment gateway to use. */
  gateway?: CreditGateway;
}

export interface CreditTopUpSuccessQuery {
  session_id: string;
}

export interface ListCreditTransactionsQuery {
  limit?: number;
  offset?: number;
}

export interface StripeCheckoutReturnQuery {
  session_id?: string;
  status?: "cancel";
}

export interface PurchaseStripePlanBody {
  /** Subscription plan identifier. */
  plan: StripePlan;
  /** Optional promotion code (case-insensitive). */
  couponCode?: string;
}

/**
 * Payments: Stripe subscriptions, Coinbase Commerce charges, credit balance and
 * top-ups, transaction history, the customer portal, and auto-recharge settings.
 */
export class PaymentsApi {
  constructor(private readonly http: HttpClient) {}

  /** Create a Coinbase Commerce charge. */
  createCoinbaseCharge<T = unknown>(
    body: CreateCoinbaseChargeBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/payments/coinbase/charge", body, options);
  }

  /** Get charge status. */
  getCoinbaseCharge<T = unknown>(
    gatewayTransactionId: string,
    query?: GetCoinbaseChargeQuery,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>(
      `/payments/coinbase/charge/${encodeURIComponent(gatewayTransactionId)}`,
      { ...options, query: { ...query, ...options?.query } },
    );
  }

  /** Get Stripe auto-recharge settings for top-up credits. */
  getAutoRecharge<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/credits/auto-recharge", options);
  }

  /** Update Stripe auto-recharge settings for top-up credits. */
  updateAutoRecharge<T = unknown>(
    body: UpdateAutoRechargeBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.patch<T>("/payments/credits/auto-recharge", body, options);
  }

  /** List saved Stripe cards for auto recharge. */
  listAutoRechargeCards<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/credits/auto-recharge/cards", options);
  }

  /** Create a Stripe SetupIntent for adding a saved card. */
  createAutoRechargeCardSetupIntent<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.post<T>(
      "/payments/credits/auto-recharge/cards/setup-intent",
      undefined,
      options,
    );
  }

  /** Update a saved Stripe card for auto recharge. */
  updateAutoRechargeCard<T = unknown>(
    paymentMethodId: string,
    body?: UpdateAutoRechargeCardBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.patch<T>(
      `/payments/credits/auto-recharge/cards/${encodeURIComponent(paymentMethodId)}`,
      body,
      options,
    );
  }

  /** Delete a saved Stripe card for auto recharge. */
  deleteAutoRechargeCard<T = unknown>(
    paymentMethodId: string,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.delete<T>(
      `/payments/credits/auto-recharge/cards/${encodeURIComponent(paymentMethodId)}`,
      options,
    );
  }

  /** Get the current user's credit balance. */
  getCreditBalance<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/credits/balance", options);
  }

  /** Create a credit top-up payment. */
  createCreditTopUp<T = unknown>(
    body: CreateCreditTopUpBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/payments/credits/top-up", body, options);
  }

  /** Handle canceled credit top-up (callback). */
  getCreditTopUpCancel<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/credits/top-up/cancel", options);
  }

  /** Handle successful credit top-up (callback). */
  getCreditTopUpSuccess<T = unknown>(
    query: CreditTopUpSuccessQuery,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>("/payments/credits/top-up/success", {
      ...options,
      query: { ...query, ...options?.query },
    });
  }

  /** Get paginated credit transaction history. */
  listCreditTransactions<T = unknown>(
    query?: ListCreditTransactionsQuery,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>("/payments/credits/transactions", {
      ...options,
      query: { ...query, ...options?.query },
    });
  }

  /** Public Stripe Checkout redirect target (browser hand-off). */
  getStripeCheckoutReturn<T = unknown>(
    query?: StripeCheckoutReturnQuery,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.get<T>("/payments/stripe/checkout/return", {
      ...options,
      query: { ...query, ...options?.query },
    });
  }

  /** Get current subscription plan for authenticated user. */
  getCurrentPlan<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/stripe/currentPlan", options);
  }

  /** Get all available subscription plans. */
  getStripePlans<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/stripe/plans", options);
  }

  /** Create a Stripe Customer Portal session. */
  createStripePortalSession<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.post<T>("/payments/stripe/portal", undefined, options);
  }

  /** Stripe Customer Portal return page. */
  getStripePortalReturn<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/payments/stripe/portal/return", options);
  }

  /** Create a Stripe Checkout Session for subscription purchase. */
  purchaseStripePlan<T = unknown>(
    body: PurchaseStripePlanBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/payments/stripe/purchasePlan", body, options);
  }
}

File: sdk/rust/tests/payments.rs
use serde_json::json;
use tinyhumans_sdk::TinyHumansClient;
use wiremock::matchers::{method, path, query_param};
use wiremock::{Mock, MockServer, ResponseTemplate};

#[tokio::test]
async fn create_coinbase_charge_posts_body() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/payments/coinbase/charge"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": {"id": "chg_1"}})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client
        .payments()
        .create_coinbase_charge(&json!({"plan": "PRO"}))
        .await
        .unwrap();

    assert_eq!(result, json!({"id": "chg_1"}));
}

#[tokio::test]
async fn get_coinbase_charge_path_and_query() {
    let server = MockServer::start().await;
    Mock::given(method("GET"))
        .and(path("/payments/coinbase/charge/tx_9"))
        .and(query_param("sync", "true"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": {"status": "PAID"}})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client
        .payments()
        .get_coinbase_charge("tx_9", &[("sync", Some("true".to_string()))])
        .await
        .unwrap();

    assert_eq!(result, json!({"status": "PAID"}));
}

#[tokio::test]
async fn get_credit_balance_unwraps_envelope() {
    let server = MockServer::start().await;
    Mock::given(method("GET"))
        .and(path("/payments/credits/balance"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": {"balanceUsd": 12.5}})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client.payments().get_credit_balance().await.unwrap();

    assert_eq!(result, json!({"balanceUsd": 12.5}));
}

#[tokio::test]
async fn update_auto_recharge_patches_body() {
    let server = MockServer::start().await;
    Mock::given(method("PATCH"))
        .and(path("/payments/credits/auto-recharge"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": {"enabled": true}})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client
        .payments()
        .update_auto_recharge(&json!({"enabled": true}))
        .await
        .unwrap();

    assert_eq!(result, json!({"enabled": true}));
}

#[tokio::test]
async fn delete_auto_recharge_card_uses_delete() {
    let server = MockServer::start().await;
    Mock::given(method("DELETE"))
        .and(path("/payments/credits/auto-recharge/cards/pm_1"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": {"deleted": true}})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client
        .payments()
        .delete_auto_recharge_card("pm_1")
        .await
        .unwrap();

    assert_eq!(result, json!({"deleted": true}));
}

#[tokio::test]
async fn create_stripe_portal_session_posts_no_body() {
    let server = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path("/payments/stripe/portal"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": {"url": "https://portal"}})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client
        .payments()
        .create_stripe_portal_session()
        .await
        .unwrap();

    assert_eq!(result, json!({"url": "https://portal"}));
}

#[tokio::test]
async fn list_credit_transactions_sends_query() {
    let server = MockServer::start().await;
    Mock::given(method("GET"))
        .and(path("/payments/credits/transactions"))
        .and(query_param("limit", "10"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"success": true, "data": []})),
        )
        .mount(&server)
        .await;

    let client = TinyHumansClient::new(server.uri());
    let result = client
        .payments()
        .list_credit_transactions(&[("limit", Some("10".to_string()))])
        .await
        .unwrap();

    assert_eq!(result, json!([]));
}
